clippy_lints/src/types.rs

```diff
@@ -583,14 +583,27 @@ fn span_precision_loss_lint(cx: &LateContext, expr: &Expr, cast_from: Ty, cast_t
 }
 
 fn span_lossless_lint(cx: &LateContext, expr: &Expr, op: &Expr, cast_from: Ty, cast_to: Ty) {
+    // The suggestion is to use a function call, so if the original expression
+    // has parens on the outside, they are no longer needed.
+    let opt = snippet_opt(cx, op.span);
+    let sugg = if let Some(ref snip) = opt {
+        if snip.starts_with('(') && snip.ends_with(')') {
```

Comment:
this may have issues with `(..) + (..)`?

Reply:
`as` happens to have greater precedence than `+` (and other operators), so as far as I can tell, that should never occur. I've now added a testcase for this.

Reply:
`(some_struct.field_of_fn_type)(foobar) as some_int` will probably still fail.

I think it might make more sense to fix this in rustc, by making sure that the removal of `ast::ExprKind::Paren` also shrinks the span.

Even rustc reports its errors including the parens: https://play.rust-lang.org/?gist=5fee704a8c23857ec818bbcc0ae7b6cb&version=nightly

Reply:
Ah, yes, that example would still fail.

I took a quick look at rustc, and it seems rustc was deliberately changed to include the parens in the span, to address https://github.com/rust-lang/rust/issues/33681.

My only other idea at this point is to rewrite the cast_lossless lint to be an EarlyLintPass instead of a late one, so that it has access to the AST nodes, including ExprKind::Paren. Would that make sense?

Reply:
We can match the expr, and only remove parens from them, and not from all expressions

Reply:
I mean only remove from binary expressions.

Reply:
It was a binary expression that motivated including the parens in the first place, in rust-lang/rust#33681, so omitting parens from binary expression spans again seems like it would re-open that issue.

Reply:
What I meant was only removing parens from binary expressions and no others.

Reply:
I'm afraid I still don't understand. rust-lang/rust#33681 wants to keep parens on binary expressions, the very place we'd remove them.

Reply:
Ignore the rust stuff. We'll fix this in clippy, by checking `expr` for binary expressions and removing the parens in that case. In all other cases just leave them there, they are rare